File: backend/app/schemas/construction.py
"""
Pydantic schemas — 构建模式（FR-012~019）

对应 api.md 第 4 节。所有请求/响应字段与 api.md 保持严格一致。
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# =============================================================================
# 启动构建 (POST /construction/start)
# =============================================================================

class StartConstructionRequest(BaseModel):
    databases: list[str] = Field(
        default=["arxiv", "openalex", "semantic_scholar"],
        description="参与检索的数据库列表",
    )
    score_threshold: int = Field(
        default=7, ge=1, le=10,
        description="有效论文评分阈值，总分≥此值视为 is_valid=True",
    )


class StartConstructionResponse(BaseModel):
    task_id: str
    project_id: str
    stage: int
    status: str
    message: str


# =============================================================================
# 构建状态查询 (GET /construction/status)
# =============================================================================

class ConstructionStatusResponse(BaseModel):
    current_stage: int | None
    stage_name: str | None
    status: str                          # idle / running / paused / error
    stage_result: dict[str, Any] | None
    started_at: datetime | None


# =============================================================================
# 检索词管理 (GET / PATCH /construction/keywords)
# =============================================================================

class KeywordItemResponse(BaseModel):
    keyword_id: str
    search_word: str
    boolean_expressions: dict[str, str] | None
    is_selected: bool
    is_searched: bool
    searched_papers_total: int


class KeywordUpdateItem(BaseModel):
    keyword_id: str | None = None        # None 表示新增；有值表示更新已有
    search_word: str
    boolean_expressions: dict[str, str] | None = None
    is_selected: bool = True


class UpdateKeywordsRequest(BaseModel):
    keywords: list[KeywordUpdateItem]


# =============================================================================
# 阶段用户操作 (POST /construction/stages/{stage}/action)
# =============================================================================

class StageActionRequest(BaseModel):
    action: str = Field(
        description="confirm=确认继续 / retry=重新执行 / skip=跳过",
    )
    # 各阶段可选修改内容（结构见 api.md §4 阶段对应的 modifications）
    modifications: dict[str, Any] | None = None


class StageActionResponse(BaseModel):
    stage: int
    action: str
    next_stage: int | None               # None 表示已是最后阶段（stage 7 完成）
    status: str
    message: str

File: frontend/api/construction.ts
import { http, tokenStore } from "./client";
import type { ISODateString, PaperSource } from "../types";

// ── 构建模式类型（对齐 api.md 第 4 节）────────────────────────────────────────

export interface StartConstructionRequest {
  databases?: PaperSource[];
  score_threshold?: number;
}

export interface StartConstructionResponse {
  task_id: string;
  project_id: string;
  stage: number;
  status: string;
  message: string;
}

export interface ConstructionStatus {
  current_stage: number;
  stage_name: string;
  status: string;
  stage_result: Record<string, unknown> | null;
  started_at: ISODateString;
}

export interface KeywordItem {
  keyword_id: string;
  search_word: string;
  boolean_expressions: {
    arxiv?: string;
    openalex?: string;
    semantic_scholar?: string;
    ads?: string;
  } | null;
  is_selected: boolean;
  searched_papers_total: number;
}

export interface KeywordUpdate {
  keyword_id?: string;      // 有 id = 更新已有；无 id = 新增
  search_word: string;
  is_selected?: boolean;
}

export interface UpdateKeywordsRequest {
  keywords: KeywordUpdate[];
}

export interface StageAction {
  action: "confirm" | "retry" | "skip";
  modifications?: Record<string, unknown>;
}

export interface StageActionResponse {
  stage: number;
  action: string;
  next_stage: number | null;
  status: string;
  message: string;
}

// ── SSE 事件数据类型 ───────────────────────────────────────────────────────────

export interface SseStageStart {
  stage: number;
  stage_name: string;
  started_at: ISODateString;
}

export interface SseStageProgress {
  stage: number;
  progress: number;
  detail: string;
}

export interface SseStagePause {
  stage: number;
  status: "paused";
  result: Record<string, unknown>;
}

export interface SseStageError {
  stage: number;
  error: string;
  retryable: boolean;
}

export interface SsePipelineComplete {
  total_papers: number;
  valid_papers: number;
  email_sent: boolean;
}

// ── API 方法 ──────────────────────────────────────────────────────────────────

export const constructionApi = {
  start: (projectId: string, body?: StartConstructionRequest) =>
    http.post<StartConstructionResponse>(
      `/projects/${projectId}/construction/start`,
      body ?? {},
    ),

  getStatus: (projectId: string) =>
    http.get<ConstructionStatus>(`/projects/${projectId}/construction/status`),

  getKeywords: (projectId: string) =>
    http.get<KeywordItem[]>(`/projects/${projectId}/construction/keywords`),

  updateKeywords: (projectId: string, body: UpdateKeywordsRequest) =>
    http.patch<KeywordItem[]>(
      `/projects/${projectId}/construction/keywords`,
      body,
    ),

  stageAction: (projectId: string, stage: number, body: StageAction) =>
    http.post<StageActionResponse>(
      `/projects/${projectId}/construction/stages/${stage}/action`,
      body,
    ),

  /**
   * 返回构建进度 SSE 流的 URL（含 token query param）。
   * 调用方自行创建 EventSource：
   *   new EventSource(constructionApi.streamUrl(projectId))
   */
  streamUrl: (projectId: string): string => {
    const token = tokenStore.get();
    const q = token ? `?token=${encodeURIComponent(token)}` : "";
    return `/api/v1/projects/${projectId}/construction/stream${q}`;
  },
};
